runtime/domain.c

```diff
@@ -821,15 +885,32 @@ CAMLexport void (*caml_leave_blocking_section_hook)(void) =
    caml_leave_blocking_section_default;
 
 CAMLexport void caml_leave_blocking_section() {
-  caml_plat_lock(&domain_self->roots_lock);
+  dom_internal* self = domain_self;
+
+  atomic_store_rel(&self->backup_thread_msg, MSG_ENTERING_OCAML);
+
+  if (self->backup_thread_running) {
+    /* Lock is necessary here to avoid races with the backup thread
```

Comment:
I didn't really understand this comment and found it confusing. 

I think what you mean is: 
 When leaving a blocking section, the backup thread will be waiting on `interruptor.lock` to service interrupts. We need to signal the backup thread such that it stops servicing interrupts and drops `domain_lock`.

Reply:
Yes. But it is a bit more subtle. Without the lock, it is not necessary that the backup thread is waiting on the `interruptor.lock`. Singaling without may lose the signal. 

Let us consider for a moment that we signal without the lock. Consider that the main thread has returned from the blocking call, and is about to call `caml_leave_blocking_section`. Consider also that the backup thread is in the inner loop, has already read the `backup_thread_msg` (which would be `MSG_IN_BLOCKING_SECTION`) but has not performed the conditional check. Now, consider that the main thread executes `caml_leave_blocking_section`. The main thread sets the backup_thread_msg to `MSG_ENTERING_OCAML` signals the condition variable (on which the backup thread is not waiting right now), and goes on to wait to obtain the `domain_lock`. 

Now, the backup thread may continue, observe that there are no interrupts to be handled, and then wait on the `interruptor.cond` condition variable. Unless another domain interrupts this domain, this domain is stuck.

It may be best to just remove that comment.

Reply:
I'm not suggesting that you don't need the lock. As you say, you need to have it so that `backup_thread_func` is forced to be at the wait or has released the lock.

Reply:
I think removing the comment is the right thing to do. It is recommended that `pthread_cond_signal` is called while holding the lock.